cmd/duffle/run.go

```diff
@@ -3,107 +3,192 @@ package main
-			target := args[0]
-			claimName := args[1]
-			storage := claimStorage()
-			c, err := storage.Read(claimName)
-			if err != nil {
-				if err == claim.ErrClaimNotFound {
-					return fmt.Errorf("Bundle installation '%s' not found", claimName)
-				}
-				return err
-			}
-
-			creds, err := loadCredentials(credentialsFiles, c.Bundle)
-			if err != nil {
-				return err
-			}
-
-			driverImpl, err := prepareDriver(driver)
-			if err != nil {
-				return err
-			}
-
-			// Override parameters only if some are set.
-			if valuesFile != "" || len(setParams) > 0 {
-				c.Parameters, err = calculateParamValues(c.Bundle, valuesFile, setParams, setFiles)
-				if err != nil {
-					return err
-				}
-			}
-
-			opRelocator, err := makeOpRelocator(relocationMapping)
-			if err != nil {
-				return err
-			}
-
-			action := &action.RunCustom{
-				Driver: driverImpl,
-				Action: target,
-			}
-
-			fmt.Fprintf(w, "Executing custom action %q for release %q", target, claimName)
-			err = action.Run(&c, creds, setOut(cmd.OutOrStdout()), opRelocator)
-			if actionDef := c.Bundle.Actions[target]; !actionDef.Modifies {
-				// Do not store a claim for non-mutating actions.
-				return err
-			}
-
-			err2 := storage.Store(c)
-			if err != nil {
-				return fmt.Errorf("run failed: %s", err)
-			}
-			return err2
+			run.action = args[0]
+			run.home = homePath()
+			return run.run()
 		},
 	}
+	run.opOutFunc = setOut(cmd.OutOrStdout())
+
 	flags := cmd.Flags()
-	flags.StringVarP(&driver, "driver", "d", "docker", "Specify a driver name")
-	flags.StringVarP(&relocationMapping, "relocation-mapping", "m", "", "Path of relocation mapping JSON file")
-	flags.StringArrayVarP(&credentialsFiles, "credentials", "c", []string{}, "Specify a set of credentials to use inside the CNAB bundle")
-	flags.StringVarP(&valuesFile, "parameters", "p", "", "Specify file containing parameters. Formats: toml, MORE SOON")
-	flags.StringArrayVarP(&setParams, "set", "s", []string{}, "Set individual parameters as NAME=VALUE pairs")
+	flags.StringVarP(&run.claimName, "claim", "i", "", "Specify the name of an existing claim (required)")
+	flags.StringVarP(&run.bundleName, "bundle", "b", "", "Specify the name of a bundle")
+	flags.StringVarP(&run.bundlePath, "bundle-is-file", "f", "", "Specify the path to a bundle.json")
+	flags.StringVarP(&run.driver, "driver", "d", "docker", "Specify a driver name")
+	flags.StringVarP(&run.relocationMapping, "relocation-mapping", "m", "", "Path of relocation mapping JSON file")
+	flags.StringArrayVarP(&run.credentialsFiles, "credentials", "c", []string{}, "Specify a set of credentials to use inside the CNAB bundle")
+	flags.StringVarP(&run.valuesFile, "parameters", "p", "", "Specify file containing parameters. Formats: toml, MORE SOON")
+	flags.StringArrayVarP(&run.setParams, "set", "s", []string{}, "Set individual parameters as NAME=VALUE pairs")
+
+	err := cmd.MarkFlagRequired("claim")
+	if err != nil {
```

Comment:
nit: in duffle, it's more usual to fold assignments and if statements together when no new variables need to last beyond the scope of the if.